### ISSUE TITLE
font display option for gantry

### ISSUE THREAD

**Issue by mikajaxxx** [gantry/gantry5#2523]
Google "page speed Insights " suggest using font-display.  but it seems gantry framework have not such option on theme.yaml

 I hope you implement this critical option for gantry for achieving better score for speed.

**Comment by thexmanxyz** [comment]
@mikajaxxx I also would love to see that implemented but if you desperately need it you can write the corresponding CSS rules yourself. Fore more information [see also](https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-display)

**Comment by mikajaxxx** [comment]
I think it could be critical. because page speed is really important nowdays, specially for google! you can't get nice score without having font-display.
Also i know how to implement this by css, but it will be much elegant if gantry itself had such option.
because "font-display" is used when you define the font so if we implement fonts by css, we loose many good thing about built in font management in gantry.
I love gantry and i used it from first release, so this is just an Opinion to make gantry even better and faster.
Thanks for your attention. 
Best :)

**Comment by thexmanxyz** [comment]
@mikajaxxx The `font-face` and `font-display` is not critical at all and that is what we are talking about here ;). Yes the page speed is important but the page speed is rated by many factors and I would count this one to the least important ones. There are factors that are more important to the overall rating. Really? I use Gantry and I get scores >92 on mobile and >99 on desktop without using `font-face` but there are still improvements possible which do not have anything to do with `font-face`.

I think you overrate the weight of that attribute a tiny bit and I assume you wouldn't get any real measurable benefits in ranking within the SERP of Google. But that is just my opinion. As I said I would also love to see that implemented but I assume the priority is not very height because the benefits are low. But you made a good point.

**Comment by thexmanxyz** [comment]
@mikajaxxx The `font-face` and `font-display` is not critical at all and that is what we are talking about here ;). Yes the page speed is important but the page speed is rated by many factors and I would count this one to the least important ones. There are factors that are more important to the overall rating. Really? I use Gantry and I get scores >92 on mobile and >99 on desktop without using `font-face` but there are still improvements possible on my side which do not have anything to do with `font-face` and this applies for all of my sites.

I think you overrate the weight of that attribute a tiny bit and I assume you wouldn't get any real measurable benefits in ranking within the SERPs of Google. But that is just my opinion. As I said I would also love to see that implemented but I assume the priority is not very height because the benefits are very low. But you made a good point. I only suggested a quick fix. If you need that ~1 point (but I have doubts that this even influences the rating at all) within page speed insight you can still implement it manually until Gantry supports it. Moreover you will not change fonts all the time. Fonts are something that is persistent on a per website base. So I see no issue in doing it manually in the meantime.

**Comment by mikajaxxx** [comment]
yeah i have 98/100 on mobile and 99/100 on desktop by many modifications, if you use 1 font with limited font-weight ... it doesn't impact so much on page load time. it is not a parameter directly score page  speed but it has significant effect on overall page loading time with modern browser. maybe because i used 3 different type of fonts with all font-weight included for two different languages. and i experienced different results.
Oh and something else maybe i should open another issue because of it /or not if you say NO :)
It will be great if we could load different fonts for every language it could help page speed . maybe it is not important too. i just tried to mention things may help gantry.
anyway thanks for your attention, and i just wanted to help improve gantry. 
Best :)

**Comment by thexmanxyz** [comment]
@mikajaxxx It's all fine I'm just saying that the influence on SERPs is probably tiny. But that is something only Google knows. But I have a strong feeling that the rating itself is the most important thing. Also Google rating factors are quite slowly adapted on new features because obviously it takes time for websites to adapt on new technologies and standards.

I have no clue if it really affects the page loading speed. AFAIK it only affects the rendering speed and rendering timing of fonts. I have not investigated or tested much but from what I read and what I understood that's the actual point of `font-display`. But maybe someone else correct me and lays down a more sophisticated explanation.

I'm just a contributor and not a repository maintainer. It sounds useful to me so go on and open another issue with that feature request. I can't decide if this is something that will be implemented in the future. The worst that can happen is that the issue will be closed or declined. 

![font-display fw](https://user-images.githubusercontent.com/17965908/61414102-767b0500-a8ed-11e9-8c3f-0ba53fb62082.png)

**Comment by Kirill-Kad** [comment]
@mikajaxxx, maybe I can help you.
> I apologize in advance for the crooked translation, I'm from Russia.
I could answer in German, but ... I use Google translator to translate everything into English. 

> And I warn you in advance. I am not a gantry 5 developer, but a regular freelancer who uses this component to write templates and managed to study it enough.

There is an alternative, in which you do not need to wait until the developers solve this issue.
To do this, create an override of the gantry5 file, which is responsible for creating the font style in your template.
Override documentation is listed here: http://docs.gantry.org/gantry5/advanced/file-overrides

In your case, you can do the following:
file <pre class="language-markup"><code>/media/gantry5/engines/nucleus/scss/vendor/bourbon/css3/_font-face.scss</code></pre>
copy and paste at
<pre class="language-markup"><code>/templates/TEMPLATE_DIR/custom/engine/scss/vendor/bourbon/css3/_font-face.scss</code></pre>

In the file that you moved to your template, you need to add the rule in @ font-face
eg:
<pre class="language-css"><code>@mixin font-face(
  $font-family,
  $file-path,
  $weight: normal,
  $style: normal,
  $asset-pipeline: $asset-pipeline,
  $file-formats: eot woff2 woff ttf svg) {

  $font-url-prefix: font-url-prefixer($asset-pipeline);

  @font-face {
    font-family: $font-family;
    font-style: $style;
    font-weight: $weight;
    font-display: swap;

    src: font-source-declaration(
      $font-family,
      $file-path,
      $asset-pipeline,
      $file-formats,
      $font-url-prefix
    );
  }
}</code></pre>
In this case, the rule will apply to all fonts except font awesome.

With fontawesome, I think you can figure it out, do the same with the above code and you will succeed.


The disadvantage of this option is that the font-display rule will be applied to all fonts that are entered in the input.fonts field.

To avoid this, I believe that it would be possible to create an override inside the template (override the file in the template to avoid problems during the update)

Theoretically, you can create an override of the template in which you can choose the individual font-display setting for each font.

In theory, this will require an override of 
<pre class="language-markup"><code>/templates/TEMPLATE_DIR/custom/blueprints/styles/font.yaml</code></pre>
in which to add the font-display parameter setting to each font, for example:

<pre class="language-markup"><code>
family-default:
      type: input.fonts
      label: Body Font
      default: 'opensans, Helvetica, Tahoma, Geneva, Arial, sans-serif'
family-default-font-display:
        type: select.selectize
        label: font-display
        default: auto
        options:
            auto: auto
            block: block
            swap: swap
            fallback: fallback
            optional: optional</code></pre>
then execute everything through mixin.

Unfortunately, I am not a connoisseur of scss, so just leave this thinking here, maybe the next commenter will know how to do it.

**Comment by mikajaxxx** [comment]
@Kirill-Kad 
Thank you very much for your Attention. I believe you are a true hero.
I will try this solution and thanks again for your advices

**Comment by 01Kuzma** [comment]
It seems that the file structure has been changed since that time. Anyway I'm unable to modify  `Helium` theme in order to add `font-display: swap;` property
